optuna/_deprecation.py

```diff
@@ -0,0 +1,103 @@
+import functools
+import inspect
+import textwrap
+from typing import Any
+from typing import Callable
+import warnings
+
+from optuna._experimental import _get_docstring_indent
+from optuna._experimental import _validate_version
+
+
+_DEPRECATION_NOTE_TEMPLATE = """
+
+.. note::
+    Deprecated in v{d_ver}. This feature will be removed in the future. The removal of this
+    feature is currently scheduled for v{r_ver}, but this schedule is subject to change.
+    See https://github.com/optuna/optuna/releases/tag/v{d_ver}.
+"""
+
+
+def deprecated(deprecated_version: str, removed_version: str, name: str = None) -> Any:
+    """Decorate class or function as deprecated.
+
+    Args:
+        deprecated_version: The version in which the target feature is deprecated.
+        removed_version: The version in which the target feature will be removed.
+        name: The name of the feature. Defaults to the function or class name. Optional.
+    """
+
+    _validate_version(deprecated_version)
+    _validate_version(removed_version)
```

Comment:
Maybe this is obvious, but do we add a check to ensure `deprecated_version < removed_version`? It may be able to prevent careless mistakes.